[FLINK-29234][runtime] JobMasterServiceLeadershipRunner handle leader event in a separate executor to avoid dead lock

reswqa opened the pull request:
## What is the purpose of the change

*Currently, in JobMasterServiceLeadershipRunner, Leader event callback such as grantLeadership and revokeLeadership will be invoked with nested lock(e.g. curator's event lock and DefaultLeaderElectionService's lock) and try to acquire JobMasterServiceLeadershipRunner's own lock. This will lead to many deadlock risks, such as the simultaneous close and grantLeadership. To avoid this, we should execute this call back in a separate thread to get rid of the nested structure of locks.*


## Brief change log

  - *JobMasterServiceLeadershipRunner handle leader event in a separate executor to avoid dead lock*


## Verifying this change

This change is covered partially by existing tests and added a new test*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

reswqa commented:
@flinkbot run azure

XComp commented:
Thanks @reswqa for this PR. I'm wondering how executing the leadership granting/revocation being called from within another thread would help fixing the issue. The locks might be still acquired concurrently in opposite orders leading to the deadlock situation.

The usecase that was described in FLINK-29234 essentially happens because the Dispatcher is stopped (which, as a consequence, would stop `JobMasterServiceLeadershipRunner`) while the `JobMasterServiceLeadershipRunner` is granted leadership causing the locks to be acquired in the opposite order.

I think the problem is that we're still trying to acquire the lock in [JobMasterServiceLeadershipRunner#runIfStateRunning:453](https://github.com/apache/flink/blob/bfe4f9cc3d67d37a2258ab4226d70b6a7d24f22c/flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java#L453) even though the `JobMasterServiceLeadershipRunner` is already switched to `STOPPED` state. I'm wondering whether we could make `JobMasterServiceLeadershipRunner#state` volatile and check the instance being in `RUNNING` state outside of the lock. But this wouldn't solve the issue entirely because there's still a slight chance that the state changes after the state check is processed but before entering the lock... :thinking:

reswqa commented:
> Thanks @reswqa for this PR. I'm wondering how executing the leadership granting/revocation call from within another thread would help fixing the issue. The locks might be still acquired concurrently in opposite orders leading to the deadlock situation.
> 
> The usecase that was described in [FLINK-29234](https://issues.apache.org/jira/browse/FLINK-29234) essentially happens because the Dispatcher is stopped (which, as a consequence, would stop `JobMasterServiceLeadershipRunner`) while the `JobMasterServiceLeadershipRunner` is granted leadership causing the locks to be acquired in the opposite order.
> 
> I think the problem is that we're still trying to acquire the lock in [JobMasterServiceLeadershipRunner#runIfStateRunning:453](https://github.com/apache/flink/blob/bfe4f9cc3d67d37a2258ab4226d70b6a7d24f22c/flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java#L453) even though the `JobMasterServiceLeadershipRunner` is already switched to `STOPPED` state. I'm wondering whether we could make `JobMasterServiceLeadershipRunner#state` volatile and check the instance being in `RUNNING` state outside of the lock. But this wouldn't solve the issue entirely because there's still a slight chance that the state changes after the state check is processed but before entering the lock... Essentially, we would need another lock guarding the state field. 🤔

Thanks @XComp for your feedback. I agree that the reason for deadlock is that the `JobMasterServiceLeadershipRunner` was granted leadership when it was being closed. 
At the beginning, I tried to make the `leaderContender.grantLeadership (issuedLeaderSessionID)` in `onGrantLeadership ` method moved outside the lock of `DefaultLeaderElectionService` to avoid nesting locks. But later, I found that there is another layer of lock nesting relationship. When executing `isLeader` and other methods, the curator framework will add a lock itself and can still reproduce the deadlock phenomenon.
So I began to consider whether there was a way to completely avoid the nesting of locks when calling the `grantLeadership` of the `JobMasterServiceLeadershipRunner`. Inspired by `ResourceManagerServiceImpl` (which executes the leader event  in a separate single thread executor), I thought that the same way should also be used to avoid the nesting of lock structures. When we let a piece of code to another thread for execution, the locks placed on the current thread will not be moved to another thread, which will enable the logic of `granteLeadership` in `JobMasterServiceLeadershipRunner` has chance to get rid of the locks  that may be imposed by the caller (whether from the leaderSelectionService or from the Curator).
We record the lock in `DefaultLeaderSelectionService` as A and the lock in `JobMasterServiceLeadershipRunner` as B.
FLINK-29234's case is as following: 
Thread t1 invoke `JobMasterServiceLeadershipRunner#closeAsync`, holds lock B, and then attempts to call `DefaultLeaderElectionService#stop` to obtain lock A.
Thread t2 invoke `DefaultLeaderElectionService#onGrantLeadership`, hold lock A, and then try to call `JobMasterServiceLeadershipRunner#grantLeadership` to obtain lock B.
After introducing an executor to process `grantLeadership`, t2 calling `onGrantLeadership` will hold lock A, move the code logic into the executor, and then release lock A without attempting to acquire lock B. It is the thread in executor's responsibility to obtain B and do the real work.
If I miss something, you can tell me. I'm willing to listen to your opinion, because I'm not particularly familiar with this part code.

reswqa commented:
@flinkbot run azure

reswqa commented:
@flinkbot run azure

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunnerTest.java:
```diff
@@ -683,21 +665,66 @@ public void testJobAlreadyDone() throws Exception {
+                };
+        contenderCloseThread.start();
+
+        // grant leadership.
+        currentLeaderDriver.isLeader();
```
This test is not necessarily failing when running it without the fix you're proposing in this PR. The `close()` call of `jobManagerRunner` needs to be halted to influence the concurrent processing of the `currentLeaderDriver.isLeader()` and the `jobManagerRunner.close()` method. Ideally, this method should fail/block consistently without the fix but should succeed with the fix.
I guess, you could achieve that by inducing your own `TestingJobMasterServiceProcess` through the `TestingJobMasterServiceProcessFactory` in [line 685](https://github.com/apache/flink/pull/21137/files#diff-c010bcd788571261ea6c43a17ad7aaea719a64a7b8a080d76939e7dbbc5a4b32R685). This `TestingJobMasterServiceProcess` gives you control over the `CompletableFuture` that's returned by [JobMasterServiceProcess.closeAsync() in JobMasterServiceLeadershipRunner.closeAsync()](https://github.com/apache/flink/blob/bfe4f9cc3d67d37a2258ab4226d70b6a7d24f22c/flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java#L145). The `JobMasterServiceLeadershipRunner` will only proceeed with stopping the `DefaultLeaderElectionService` if this `CompletableFuture` completes. WDYT?

reswqa replied:
@XComp Thank you for your advice. 
At the beginning, I didn't think of a good way to control the completion of completableFuture, and the probability of deadlock reproduce in my local environment is very high. But your suggestion did find a good control injection point, so i decided to test it in this way. 
But if we want to reproduce the problem 100%, it seems that we need another control point. Maybe `TestingLeaderElectionDriver#isLeader` is a suitable choice. What's your opinion?

reswqa replied:
@XComp Thank you for your advice.
At the beginning, I didn't think of a good way to control the completion of completableFuture, and the probability of deadlock reproduce in my local environment is very high. But your suggestion did find a good control injection point, so i decided to test it in this way.
But if we want to reproduce the problem 100%, it seems that we need another control point. Maybe TestingLeaderElectionDriver#isLeader is a suitable choice. What's your opinion?

XComp replied:
I went ahead and created a PR with a running example. See reswqa/flink#1 and feel free to edit it. Please see the PRs description. It's not meant to be merged as is but rather an example on how you would be able to test it. I'm happy to discuss any questions you have. But I hope the inline comments are helpful already.

XComp reviewed (commented):
Thanks for the thorough explanation. Initially, I missed the point that we're only moving the nested lock to a separate thread. I walked through the process once more this morning and it makes sense.

@zentol may you double-check the reasoning to be on the safe side?

XComp commented:
We might also want to check the other `LeaderContender` implementations (as you already mentioned, `ResourceManagerImpl` does not suffer from this bug but the others might because they don't use this `handleLeaderEventHandler` pattern.

reswqa commented:
> We might also want to check the other `LeaderContender` implementations (as you already mentioned, `ResourceManagerImpl` does not suffer from this bug but the others might because they don't use this `handleLeaderEventHandler` pattern.

@XComp I totally agree that we should also check the other `LeaderContender` to ensure they are not suffer from this. However, from my point of view, in consideration of limiting the scope of this pull request, I suggest that only the `JobMasterServiceLeadershipRunner` be tested in this PR. For other `LeaderContender`, I will open a separate ticket for tracking. WDYT?

reswqa commented:
@XComp About your [proposal](https://github.com/reswqa/flink/pull/1) , I add a very short sleep after `triggerClassLoaderLeaseRelease.await();`, the deadlock problem can be 100% reproduced on my own machine. Although theoretically there is still the possibility of non recurrence, but I am sure that this short sleep can almost absolutely trigger `grantLeadership` before `running` state becomes false. 
Personally, I think it is acceptable to introduce a very short sleep (I used 5ms) to fight for the deadlock problem to recur almost every time, and it will not significantly increase the running time of CI. Of course, if we can think of a better way to solve this problem, I will be more willing to accept it.
In addition, I found that if sleep is not introduced, run the whole test class on my and my colleagues' machines, the deadlock problem can hardly be reproduced. However, if run `testJobMasterServiceLeadershipRunnerCloseWhenElectionServiceGrantLeaderShip` separately, the probability of deadlock will be greatly increased. I don't have a deep understanding of Junit's testing framework, so I don't know why there is such a difference. You can try to run the entire test class directly to help you reproduce the problem.

XComp commented:
> @XComp I totally agree that we should also check the other LeaderContender to ensure they are not suffer from this. However, from my point of view, in consideration of limiting the scope of this pull request, I suggest that only the JobMasterServiceLeadershipRunner be tested in this PR. For other LeaderContender, I will open a separate ticket for tracking. WDYT?

Yup, I agree. Creating follow-up Jira issues makes sense.

> testJobMasterServiceLeadershipRunnerCloseWhenElectionServiceGrantLeaderShip separately, the probability of deadlock will be greatly increased. I don't have a deep understanding of Junit's testing framework, so I don't know why there is such a difference. Maybe you can try to run the entire test class directly to help you reproduce the problem.

Thanks for the hint. :+1:

XComp commented:
Thanks for your patience on that PR. I gave it another thought: You're right about the issue with the 5ms delay being necessary to enter the synchronized block of the `grantLeadership` call hierarchy. The issue is that we don't have any means of verify that we're in the synchronized block of `DefaultLeaderElectionService#lock` in `DefaultLeaderElectionService#onGrantLeadership` but we trigger this implicly through `TestingLeaderElectionDriver#isLeader()`. This applies to your and my proposal and, therefore, the delay (i.e. the 5ms) are actually necessary. I'm not a big fan of these 5ms because they tend to make these tests flaky. We should be really clear in an inline comment on why this is necessary.

That said, I tend to prefer my proposal, because it keeps the test-specific logic in the test method. Your proposal adds a test-specific callback to `TestingLeaderElectionDriver` which is some kind of generic test utility class. Mixing test utility classes with actual test implementations messes things up in the long run usually. That's the reason we started to come up with `Testing*` implementations like I proposed with `TestingJobMasterServiceProcess2` in my PR. WDYT?

reswqa commented:
@XComp Thank you very much for your suggestions and review. I totally agree with you that introduction of such callbacks in the `TestingLeaderSelectionDriver` will make things a bit troublesome and strange. So I decided to take your proposal. At the same time, I refactor the 'TestingJobMasterServiceProcess', and introduced `Thread.sleep (5ms)` to your proposal and added the corresponding explanation. By the way, I added you as a co-author to thank you for your proposal.

zentol commented:
I'm not really sold on the idea of solving a deadlock by introducing yet another thread into the mix.

To me the issue stems more from both the runner and election service calling into each other under locks, and locks maybe being way too broad.
For example, why is `Runner#closeAsync` doing the entire shutdown under the lock? Modifying the state _should_ suffice, because all other operations are checking that it's running.

reswqa commented:
@zentol Thank you for your suggestion. To be honest, I admit that your proposal is feasible. At first, I considered similar changes, but the main reasons why I haven't done in that way are as follows:
1. The leaderEvent (e.g. `grantLeadership`/`revokeLeadership`) method often carries the lock of multi-layer nested structure, which is easy to cause deadlock. Even if we can guarantee that we will not access these nested lock while holding the runner's own lock, some future changes may still be inadvertently troubled by this, and these deadlock problems are not easy to be found in time during the review.
2. `ResourceManagerServiceImpl` uses the same method (a separate thread handles the leaderEvent), which is conducive to the unification of handle leader event among `leaderContender`s.
3. Even if we solve the problem that the lock in `closeAsync` is too broad, there may be similar problems in other methods. Even if there is no such problem now, it is difficult to ensure that no one will fall into the trap later.
Personally, in the long run, avoiding nested lock structures can avoid similar potential deadlock problems almost once and for all. So, I think it's worth paying for it. Of course, if you think this way is unacceptable, I am not unshakable.

XComp commented:
> To me the issue stems more from both the runner and election service calling into each other under locks (== fundamental issue that should never happen), and locks maybe being way too broad.
For example, why is Runner#closeAsync doing the entire shutdown under the lock? Modifying the state should suffice, because all other operations are checking that it's running.

As far as we concluded, the problem appears if the `CompletableFuture` that's returned by `JobMasterServiceProcess#closeAsync` (see [JobMasterServiceLeadershipRunner.java:145](https://github.com/apache/flink/blob/113299701cc0c41bf7fc4bbe86cebd3beea8dbe3/flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java#L145)) completes before triggering the callback that follows (that releases the `ClassLoaderLease` and stops the `DefaultLeaderElectionService`). In that case, the entire callback will be executed in the synchronized block rightaway, which is not what we want. We could work around that issue by calling `runAfterwardsAsync`, instead, which would make sure that the followup calls are not executed within the synchronized block. It feels like a similar pattern (i.e. introducing an async execution in a separate thread) to what we have in the PR right now (just in a different location). Does that sound more reasonable?

xintongsong commented:
Thanks @reswqa for the fix, and thanks all for the discussion.

For this specific case, I think it indeed can be fixed either way, introducing a thread or making sure `leaderElectionService.stop()` is called outside the lock scope. However, from the perspective of maintenance, I think having a dedicated `handleLeaderEventExecutor` is easier to understand and maintain compared to carefully decide which lines of code needs to be executed under lock and which needn't. Admittedly, it comes with a price of an extra thread, which is probably fine in this one-per-contender case.

So +1 for the current approach. The PR LGTM.

reswqa commented:
@flinkbot run azure

reswqa commented:
@flinkbot run azure

zentol commented:
I don't see how this is supposed to be good from a maintenance perspective.

If you look at the code without context you now have 2 arbitrary calls go through an executor and you make similar decisions as with locking ("what operations should run in the executor").
The javadocs states that all operations are serialized, yet we need an executor for certain operations to make things work. That's just an accident waiting to happen.

Additionally the proposal changes the behavior slightly, since calling grant/revokeLeadership no longer has an immediate effect (as it now effectively runs in the background). If this is ok (not sure if it is), then I'd rather see us pushing more code into the `sequentialOperation` and always running that in the executor. Then you at least have a consistent behavior.

zentol commented:
> In that case, the entire callback will be executed in the synchronized block rightaway, which is not what we want.

Which implies that the future calls should never run under the lock in the first place. So let's not put them into the synchronized block.

zentol reviewed (changes requested):
Just to avoid misunderstandings, I'm -1 on this change as is.

xintongsong commented:
> If you look at the code without context you now have 2 arbitrary calls go through an executor and you make similar decisions as with locking ("what operations should run in the executor").

It would be ideal if `JobMasterServiceLeadershipRunner` need not be aware of the lock inside the leader election service. Unfortunately, I don't see a good way to do so.

In the current approach, the name of the executor clearly states that it is supposed to be used for handling the leadership related events, which are granting and revoking. You can safely access the runner lock in any other methods of the runner, without worrying about causing deadlocks with another lock from another class. Otherwise, imagine you called a method `a()` in `grantLeadership()`, and `b()` in `a()`, and `c()` in `b()`, etc. One can easily get confused whether `c()` is called under the leader election service lock and whether it is safe to access the runner lock. That is to explain why I think the current approach is good for the maintenance. 

>The javadocs states that all operations are serialized, yet we need an executor for certain operations to make things work. That's just an accident waiting to happen.

I don't get what do you mean by an accident waiting to happen. The JavaDoc says the granting and revoking of leadership are serialized, not the other public methods of the class.

> calling grant/revokeLeadership no longer has an immediate effect (as it now effectively runs in the background)

Exactly. The change is based on the fact that the runner neither need to do the works synchronously in grant/revokeLeadership, nor need/want to be aware of the lock that the leader election service put upon the two methods.

> I'd rather see us pushing more code into the sequentialOperation and always running that in the executor

Well, that should also work. I don't see much differences compare the current approach, except for more code changes. But I'm not opposed to it.

XComp reviewed (commented):
Thanks for the interesting discussion on that topic. To be honest, I didn't have a strong opinion on having the extra thread but I saw benefits on moving the leadership handling into its own event thread on the `DefaultLeaderElectionService` side to ensure that we prevent similar situations to arise in other locations where the leader election is used (similarly to how it's done for the `ResourceManagerImpl` leader election).

But I see @zentol's point (AFAIU) that we're working around an actual issue on the `JobMasterServiceLeadershipRunner` side: In the end, aren't we using the `LeaderElectionService` wrongly by only calling `JobMasterServiceLeadershipRunner#grantLeadership` and `JobMasterServiceLeadershipRunner#revokeLeadership` sequentially. Both methods are used for managing the leadership state. But the same applies to the `JobMasterServiceLeadershipRunner#closeAsync`, it influences the leadership by stopping the service and, therefore, implictly revoking leadership. So, from a conceptual point of view, one could argue that `closeAsync` should be chained up in the `sequentialOperations` as well as proposed by @zentol.

@reswqa I added some minor mostly cosmetic comments to the PR as it is right now.

zentol commented:
> Well, that should also work. I don't see much differences compared to the current approach, except for more code changes. But I'm not opposed to it.

One advantage of my proposal is that you can't run into TOCTOU races. If the state check runs as part of the sequential operation we are sure that it only runs if we're actually still running by the time the executor processes is.
As is, you check that its currently running, schedule an operation, but by the time the operation runs we may have already shut down.

The second main defect is that some calls prepared in closeAsync may _or may not_ run under the lock. That's something we must fix in any case (and is the actual bug fix!); the introduction of the executor could be a means to that end, or just be a refactoring.

reswqa commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunnerTest.java:
```diff
@@ -46,45 +49,36 @@
 
 /** Tests for the {@link JobMasterServiceLeadershipRunner}. */
-public class JobMasterServiceLeadershipRunnerTest extends TestLogger {
+@ExtendWith(TestLoggerExtension.class)
```
We already have src/test/resources/META-INF/services/org.junit.jupiter.api.extension.Extension in flink-runtime module.

reswqa replied:
Ahh, Thank you for your clarification, added this annotation accidentally~

xintongsong commented:
This is indeed a fascinating discussion.

I see the point that there are some calls in `cloaseAsync` that may or may not run under the lock. I also agree that it would be nice the lock only covers the calls that need to be guarded and leave the unnecessary calls outside.

I think my focus is more about preventing (or at least reducing the chances of) similar problems. In that sense, it's probably better that the runner does not need to (or as least as possible) worry about another lock from the leader election service. To that end, I think @XComp's proposal makes more sense than mine. Ideally, leader election service should not call `contender.grant/revokeLeadership()` under a lock while the same lock can be accessed by the contender via `stop()`.

To clarify, fixing the problem on the service side does not justify that calls in `cloaseAsync` may or may not run under the lock. Ideally, `A` should not call `B.b()` under a lock while another method `A.a()` called by `B` can access the same lock. That applies to both the service and the runner, and maybe we should fix that for both.

All in all, I'd suggest:
- Fix the runner side issue in `closeAsync` in this PR. For this purpose solely, we probably don't need a dedicated executor.
- Fix the leader election service side issue with a dedicated executor, in a follow up ticket. That would affect all contenders and we need to carefully check that no existing contenders are relying on the current behavior that `grant/revokeLeadership` are called under lock. We should also clean up things like `ResourceManagerServiceImpl.handleLeaderEventExecutor`.

WDYT?

reswqa commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java:
```diff
@@ -151,10 +155,12 @@ public CompletableFuture<Void> closeAsync() {
-                                LOG.debug(
-                                        "Leadership runner for job {} has been terminated.",
-                                        getJobID()));
+                        (unused, throwable) -> {
+                            LOG.debug(
+                                    "Leadership runner for job {} has been terminated.",
+                                    getJobID());
+                            handleLeaderEventExecutor.shutdown();
+                        });
```
good idea.

reswqa commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunnerTest.java:
```diff
@@ -93,48 +87,47 @@ public class JobMasterServiceLeadershipRunnerTest extends TestLogger {
-            assertThat(
-                    jobManagerRunner.getJobMasterGateway(),
-                    FlinkMatchers.futureWillCompleteExceptionally(Duration.ofMillis(5L)));
+            assertThat(jobManagerRunner.getJobMasterGateway())
+                    .failsWithin(5L, TimeUnit.MILLISECONDS);
```
Thank you for reminding me about this. If I understand correctly, can I change this timeout assertion to the approach like `assertThatThrownBy(() -> jobManagerRunner.getJobMasterGateway().get())` to rely on Azure's global timeout in this way?
As for the migration of JUnit5, I saw this prompt: "Unless there is a specific reason, make sure you use JUnit 5 and AssertJ when contributing to Flink with new tests and even when modifying existing tests."  in the `Apache Flink Code Style and Quality Guide`. From my perspective, it seems a bit bad to have JUnt4 and JUnit5 dependencies in the test class at the same time, so I think this may be necessary.

reswqa replied:
BTW, I have not removed the timeout in the test at present, because according to our discussion, the production code maybe should change, and I will modify it together after we reach an agreement on the implementation.

reswqa commented:
Thanks everyone for your comments and discussion, which is very meaningful to me. Based on the above discussion, I think @xintongsong's suggestion looks good to me, +1 for this plan.
We should first deal with the problem of too broad lock(the real reason of this ticket) at `JobMasterServiceLeadershipRunner#closeAsync`, which should be consistent with @zentol 's opinion. At the same time, I still think we should be committed to preventing the same thing from happening, so I think the second proposal is also very make sense.  WDYT?

zentol commented:
Sounds good to me.

xintongsong reviewed (approved):
Thanks @reswqa. LGTM.

@XComp @zentol Any further concerns? Or I'll merge this.

xintongsong commented:
@reswqa, as previously discussed, please remember to open a ticket for fixing the leader election service side and all other contenders.
